### PR TITLE
introducing realtions between corpus assets

### LINKED ISSUE #2: Ensure that utterance exists when creating label lists

### PR DESCRIPTION
**Pull request by ynop**
Fixes #2, since label-lists are now directly assigned to utterances.

### CODE REVIEW

File: pingu/corpus/assets/label.py [hunk 1]
```diff
@@ -13,13 +13,15 @@ class Label(object):
...
+        label_list (LabelList): The label-list this label is belonging to.
     """
-    __slots__ = ['value', 'start', 'end']
+    __slots__ = ['value', 'start', 'end', 'label_list']
 
     def __init__(self, value, start=0, end=-1):
         self.value = value
         self.start = start
         self.end = end
+        self.label_list = None
```
Comment (aahlenst) [inline pingu/corpus/assets/label.py]: `label_list=None` as argument would make sense, no?

File: pingu/corpus/assets/label.py [hunk 2]
```diff
@@ -33,6 +35,26 @@ def __lt__(self, other):
     def __repr__(self) -> str:
         return 'Label({}, {}, {})'.format(self.value, self.start, self.end)
 
+    def read_samples(self, sr=None):
+        """
+        Read the samples of the utterance.
+
+        Args:
+            sr (int): If None uses the sampling rate given by the file, otherwise resamples to the given sampling rate.
```
Comment (aahlenst) [inline pingu/corpus/assets/label.py]: What happens if sr is bigger than the sampling rate of the file? Exception?

File: pingu/corpus/assets/label.py [hunk 3]
```diff
@@ -54,11 +77,14 @@ class LabelList(object):
...
-    __slots__ = ['labels', 'idx']
+    __slots__ = ['idx', 'labels', 'utterance']
 
     def __init__(self, idx='default', labels=[]):
         self.idx = idx
-        self.labels = list(labels)
+        self.utterance = None
+
+        self.labels = []
+        self.extend(labels)
```
Comment (aahlenst) [inline pingu/corpus/assets/label.py]: `labels=None` in the argument list and `self.labels = labels or []` should save us a loop over the list of labels.

File: pingu/corpus/assets/label.py [hunk 4]
```diff
@@ -76,7 +103,8 @@ def extend(self, labels):
         Args:
             labels (list): Labels to add.
         """
-        self.labels.extend(labels)
+        for label in labels:
+            self.append(label)
```
Comment (aahlenst) [inline pingu/corpus/assets/label.py]: Why `append()` instead of `extend()`?

File: pingu/corpus/assets/label.py [hunk 5]
```diff
@@ -13,13 +13,15 @@ class Label(object):
...
+    __slots__ = ['value', 'start', 'end', 'label_list']
 
     def __init__(self, value, start=0, end=-1):
         self.value = value
         self.start = start
         self.end = end
+        self.label_list = None
 
     def __eq__(self, other):
         return (self.start, self.end, self.value.lower()) == (other.start, other.end, other.value.lower())
```
Comment (aahlenst) [inline pingu/corpus/assets/label.py]: The test should make clear that we're ignoring the referenced label list on purpose and that two labels are considered the same even the label lists are different. Same applies to `__lt__`.

File: pingu/corpus/assets/label.py (hunk 3 above)
Comment (aahlenst) [reply pingu/corpus/assets/label.py]: Considering that the other constructors (`Utterance`, …) automatically set the references to the owned objects during construction, this should probably happen here too.

File: pingu/corpus/assets/utterance.py [hunk 6]
```diff
@@ -1,21 +1,110 @@
...
-    def __init__(self, idx, file_idx, issuer_idx=None, start=0, end=-1):
+    def __init__(self, idx, file, issuer=None, start=0, end=-1):
         self.idx = idx
-        self.file_idx = file_idx
-        self.issuer_idx = issuer_idx
+        self.file = file
+        self.issuer = issuer
         self.start = start
         self.end = end
+        self.label_lists = {}
```
Comment (aahlenst) [inline pingu/corpus/assets/utterance.py]: Argument?

File: tests/corpus/assets/test_file.py [hunk 7]
```diff
@@ -36,16 +37,10 @@ def test_duration(self):
...
-        ref = np.array([7.62939453e-05, -3.05175781e-05, -6.10351562e-05,
-                        3.05175781e-05, -9.15527344e-05, 3.05175781e-05,
-                        1.52587891e-05, 1.83105469e-04, -1.06811523e-04,
-                        -3.05175781e-05, -1.67846680e-04, -6.10351562e-05,
-                        1.37329102e-04], dtype=np.float32)
-
-        assert np.array_equal(ref, self.stereo_file.read_samples())
+        expected, __ = librosa.core.load(STEREO_22050_32BIT_13, sr=None)
+        actual = self.stereo_file.read_samples()
+        assert np.array_equal(actual, expected)
```
Comment (aahlenst) [inline tests/corpus/assets/test_file.py]: A resampling test (`sr != None`) would be nice, too.

File: pingu/corpus/assets/label.py (hunk 1 above)
Comment (ynop) [reply pingu/corpus/assets/label.py]: The ``label_list`` attribute is set automatically, when the label is added to a label-list (in the append method). Therefore i think it is not useful.

File: pingu/corpus/assets/label.py (hunk 2 above)
Comment (ynop) [reply pingu/corpus/assets/label.py]: Librosa returns an upsampled signal.

File: pingu/corpus/assets/label.py (hunk 3 above)
Comment (ynop) [reply pingu/corpus/assets/label.py]: Relation is set in ``append``. That is why i do all additions, by calling ``append()``.

File: pingu/corpus/assets/utterance.py (hunk 6 above)
Comment (ynop) [reply pingu/corpus/assets/utterance.py]: I will do that.

File: tests/corpus/assets/test_file.py (hunk 7 above)
Comment (ynop) [reply tests/corpus/assets/test_file.py]: yup

**Pull request merged**